src/charm.py

```diff
@@ -0,0 +1,595 @@
+import shutil
+import urllib.error
+from pathlib import Path
+from typing import TYPE_CHECKING, Callable, Dict, Optional, TypeVar
+
+import jinja2
+from ops.charm import (
+    ActionEvent,
+    CharmBase,
+    ConfigChangedEvent,
+    InstallEvent,
+    StopEvent,
+    UpgradeCharmEvent,
+)
+from ops.framework import EventBase, StoredState
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+
+from errors import RunnerError, SubprocessError
+from event_timer import EventTimer, TimerDisableError, TimerEnableError
+from github_type import GitHubRunnerStatus
+from runner_manager import RunnerManager, RunnerManagerConfig
+from runner_type import GitHubOrg, GitHubRepo, ProxySetting, VirtualMachineResources
+from utilities import execute_command, get_env_var, retry
+
+if TYPE_CHECKING:
+    from ops.model import JsonObject  # pragma: no cover
+
+logger = logging.getLogger(__name__)
+
+
+class ReconcileRunnersEvent(EventBase):
+    """Event representing a periodic check to ensure runners are ok."""
+
+
+class UpdateRunnerBinEvent(EventBase):
+    """Event representing a periodic check for new versions of the runner binary."""
+
+
+CharmT = TypeVar("CharmT")
+EventT = TypeVar("EventT")
+
+
+def catch_unexpected_charm_errors(
+    func: Callable[[CharmT, EventT], None]
+) -> Callable[[CharmT, EventT], None]:
+    """Catch unexpected errors in charm.
+
+    This decorator is for unrecoverable errors and sets the charm to
+    `BlockedStatus`.
+
+    Args:
+        func: Charm function to be decorated.
+
+    Returns:
+        Decorated charm function with catching unexpected errors.
+    """
+
+    @functools.wraps(func)
+    def func_with_catch_unexpected_errors(self, event: EventT) -> None:
+        # Safe guard against unexpected error.
+        try:
+            func(self, event)
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            logger.exception(err)
+            self.unit.status = BlockedStatus(str(err))
+
+    return func_with_catch_unexpected_errors
+
+
+def catch_unexpected_action_errors(
+    func: Callable[[CharmT, ActionEvent], None]
+) -> Callable[[CharmT, ActionEvent], None]:
+    """Catch unexpected errors in actions.
+
+    Args:
+        func: Action function to be decorated.
+
+    Returns:
+        Decorated charm function with catching unexpected errors.
+    """
+
+    @functools.wraps(func)
+    def func_with_catch_unexpected_errors(self, event: ActionEvent) -> None:
+        # Safe guard against unexpected error.
+        try:
+            func(self, event)
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            logger.exception(err)
+            event.fail(f"Failed to get runner info: {err}")
+
+    return func_with_catch_unexpected_errors
+
+
+class GithubRunnerCharm(CharmBase):
+    """Charm for managing GitHub self-hosted runners."""
+
+    _stored = StoredState()
+
+    service_token_path = Path("service_token")
```

Comment:
It is probably better to move these to a separate module and declare them as constants

Reply:
Minor comments